File: 백준/Gold/17370. 육각형 우리 속의 개미/육각형 우리 속의 개미.py

dy = [0,1,1,0,-1,-1]
dx = [1,0,-1,-1,0,1]

def solve(y, x, dir, cnt):
    ny = y + dy[dir]
    nx = x + dx[dir]

    if visited[ny][nx]:
        if cnt == N:
            return 1
        return 0

    if cnt == N:
        return 0

    ret = 0
    visited[ny][nx] = True
    ret += solve(ny,nx, (dir + 1) % 6, cnt + 1)
    ret += solve(ny, nx, (dir + 5) % 6, cnt + 1)
    visited[ny][nx] = False
    return ret

if __name__ == "__main__":
    N = int(input())

    MAX_R = 50;
    visited = [[False for _ in range(MAX_R+1)] for _ in range(MAX_R + 1)]
    sy = MAX_R // 2
    sx = MAX_R // 2
    visited[sy][sx] = True

    print(solve(sy,sx,0,0))

File: 백준/Gold/17370. 육각형 우리 속의 개미/육각형 우리 속의 개미.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		
		visited = new boolean[MAX_R+1][MAX_R+1];
		
		int sy = MAX_R/2;
		int sx = MAX_R/2;
		visited[sy][sx] = true;
		
		solve(sy,sx,0,0);
		System.out.println(ans);
	}
	
	
	static int dy[] = {1,2,1,-1,-2,-1};
	static int dx[] = {1,0,-1,-1,0,1};
	static void solve(int y,int x ,int dir, int cnt) {
		int ny = y + dy[dir];
		int nx = x + dx[dir];
		
		if(visited[ny][nx]) {
			if(cnt == N)
				ans++;
			return;
		}
		
		if(cnt == N)
			return;
		
		visited[ny][nx] = true;
		solve(ny,nx,(dir + 1) % 6, cnt + 1);
		solve(ny,nx, (dir + 5)% 6, cnt + 1);
		visited[ny][nx] = false;
	}
	
	
	static int ans;
	static int N;
	static int MAX_R = 100;
	static boolean visited[][];
	
}
